core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -142,26 +182,103 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable Aggr
-        for (int i = 0; i < this.parameters.size(); ++i) {
-            Parameter entry = this.parameters.get(i);
-            if (entry.isPathParam()) {
-                String value = ctx.pathParam(entry.getName());
-                assert value != null;
-                parameters[i] = convertParameter(value, entry.getType());
-            } else if (entry.getType().isAssignableFrom(ServiceRequestContext.class)) {
-                parameters[i] = ctx;
-            } else if (entry.getType().isAssignableFrom(HttpRequest.class)) {
-                parameters[i] = req;
-            } else if (entry.getType().isAssignableFrom(AggregatedHttpMessage.class)) {
-                parameters[i] = message;
+        ParamMap paramMap = null;
+        Object[] values = new Object[parameters.size()];
+        for (int i = 0; i < parameters.size(); ++i) {
+            Parameter entry = parameters.get(i);
+            final String value;
+            switch (entry.parameterType()) {
+                case PATH_PARAM:
+                    value = ctx.pathParam(entry.name());
+                    assert value != null;
+                    values[i] = convertParameter(value, entry.type());
+                    break;
+                case PARAM:
+                    if (paramMap == null) {
+                        paramMap = paramMapOf(ctx, req, message);
+                    }
+                    value = getParam(paramMap, entry);
+                    values[i] = value != null ? convertParameter(value, entry.type()) : null;
+                    break;
+                case PREDEFINED_TYPE:
+                    if (entry.type().isAssignableFrom(ServiceRequestContext.class)) {
+                        values[i] = ctx;
+                    } else if (entry.type().isAssignableFrom(HttpRequest.class)) {
+                        values[i] = req;
+                    } else if (entry.type().isAssignableFrom(AggregatedHttpMessage.class)) {
+                        values[i] = message;
+                    } else if (entry.type().isAssignableFrom(ParamMap.class)) {
+                        if (paramMap == null) {
+                            paramMap = paramMapOf(ctx, req, message);
+                        }
+                        values[i] = paramMap;
+                    }
+                    break;
             }
         }
-        return parameters;
+        return values;
+    }
+
+    /**
+     * Returns a map of parameters decoded from a request.
+     */
+    private ParamMap paramMapOf(ServiceRequestContext ctx, HttpRequest req,
+                                @Nullable AggregatedHttpMessage message) {
+        try {
+            if (req.method() == HttpMethod.GET) {
+                return wrap(new QueryStringDecoder(ctx.query(), false).parameters());
+            }
+            if (hasFormData(req)) {
+                assert message != null;
+                String body = message.content().toStringUtf8();
```

Comment:
Actually, we have to respect the charset while decoding the form. I think we can use `toStringAscii()` here.